framework: Add linux platform DB

simonbaren opened the pull request:
Currently the linux BPL is more or less a dummy BPL just returns constant values.
For more flexibility and further progress it was needed to add a DB that can be changed for each agent that we want to run.

simonbaren committed:
bpl: cmake: Add platform_db file and install it

The platform_db is used to read platform setting.
It is used only for linux BPL (other platform have their own DB).

Signed-off-by: Semyon Barenboym <simonbaren@gmail.com>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

simonbaren committed:
framework: utils: add string trimming utilities

Add string trimming utility to bpl common, which are needed for
parsing the platform DB file in case of Linux platform configuration
where we don't have UCI.

Signed-off-by: Semyon Barenboym <simonbaren@gmail.com>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

simonbaren committed:
bpl: linux: get master and operating mode from DB

Add the functions that actually read from the platform_db file.
The function expects that the db will look like: "some_param=value"

Read is_master and operating mode from DB instead of returning dummy
values.

Signed-off-by: Semyon Barenboym <simonbaren@gmail.com>
Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

simonbaren committed:
bpl: linux: refactor return values

use predefined return values instead of magic numbers (-1/0/1)

Signed-off-by: Semyon Barenboym <simonbaren@gmail.com>

tomereli commented on framework/platform/bpl/common/utils/utils.cpp:
```diff
@@ -24,6 +24,24 @@ namespace beerocks {
 namespace bpl {
 namespace utils {
 
+static const std::string WHITESPACE_CHARS(" \t\n\r\f\v");
```
commit message missing sign off by.
Also, I would change the commit message to 

> Add string trimming utility to bpl common. These are needed for parsing the platform DB file in case of Linux platform.

tomereli replied:
Also since we're going to need this also for BWL, isn't it more proper to add all the utils to mapfcommon.so so it can be reused?

simonbaren replied:
for BWL there is beerocks::string_utils

arnout replied:
IMO all of common/beerocks should move into framework, and bcl should be merged into mapf::common...

But for now, the current situation is :ok:

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,8 +15,58 @@ using namespace mapf;
 
```
Again, commit message missing sign off by.

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,8 +15,58 @@ using namespace mapf;
 
 using namespace beerocks::bpl;
 
+#define RETURN_OK 0
```
don't use RETURN_OK and RETURN_ERR - this file by default returns 0 for success and 1 otherwise. You can return 0 for success and -1 otherwise, will work just fine as well.

simonbaren replied:
0 for success and -1 for error is the convention in the other BPL's. and since they are replaceable the convention must stay.
And I'd rather have one place to change those return values instead of going over each API...

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,8 +15,58 @@ using namespace mapf;
 
+int bpl_cfg_get_param(const std::string &param, std::string &value) 
+{
+    MAPF_ERR("file=" << PLATFORM_DB_PATH << std::endl);
```
1. Remove error print - in any case it is debug.
2. You should add `#ifndef` to make sure `PLATFORM_DB_PATH is defined (to the top of this file or a header which everyone uses):
```
#ifndef PLATFORM_DB_PATH
#error "PLATFORM_DB_PATH not defined!"
#endif
```

simonbaren replied:
Ok

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,8 +15,58 @@ using namespace mapf;
+
+int bpl_cfg_get_param_int(const std::string &param, int *value)
+{
+    std::string str_value;
```
Need to check that value is not `nullptr`. Also, why not use a reference instead of a pointer?

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
 
 int bpl_cfg_is_enabled() { return 1; }
 
-int bpl_cfg_is_master() { return 1; }
```
Again no sign off by

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
+        return RETURN_ERR;
+    } 
+
+    if (mode_str == "Multi-AP-Controller-and-Agent") return 1;
```
please add line break before return statements

tomereli replied:
```suggestion
    return (mode_str == "Multi-AP-Controller-and-Agent") ? 1 : 0;
```

simonbaren replied:
I leave that open for more modes. 
Also, your suggestion does not cover wrong configuration

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
+    if (mode_str == "Multi-AP-Controller-and-Agent") return 1;
+    else if (mode_str == "Multi-AP-Agent")return 0;
+    
+    MAPF_ERR("bpl_cfg_is_master: Unexpected ManagementMode " << mode_str << std::endl);
```
No need for `std::endl` - this is handled by the logger.

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
+int bpl_cfg_is_master() { 
+    std::string mode_str;
+    if (bpl_cfg_get_param("management_mode=", mode_str) < 0) {
+        MAPF_ERR("bpl_cfg_is_master: Failed to read ManagementMode\n");
```
No need for `\n`

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
-int bpl_cfg_get_operating_mode() { return BPL_OPER_MODE_GATEWAY; }
+int bpl_cfg_get_operating_mode() 
+{ 
+    int retVal                            = 0;
```
why so many spaces?

simonbaren replied:
clang stuff from different file....

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
+    int retVal                            = 0;
+    std::string op_mode;
+    if (bpl_cfg_get_param("operating_mode=", op_mode) < 0) {
+        MAPF_ERR("bpl_cfg_get_operating_mode: Failed to read OperatingMode\n");
```
\n not needed

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
+    std::string op_mode;
+    if (bpl_cfg_get_param("operating_mode=", op_mode) < 0) {
+        MAPF_ERR("bpl_cfg_get_operating_mode: Failed to read OperatingMode\n");
+        retVal = -1;
```
just return -1

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
+    if (bpl_cfg_get_param("operating_mode=", op_mode) < 0) {
+        MAPF_ERR("bpl_cfg_get_operating_mode: Failed to read OperatingMode\n");
+        retVal = -1;
+    } else {
```
no need for else if you return above

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -69,9 +69,46 @@ int bpl_cfg_get_param_int(const std::string &param, int *value)
+        MAPF_ERR("bpl_cfg_get_operating_mode: Failed to read OperatingMode\n");
+        retVal = -1;
+    } else {
+        if (op_mode == "Gateway") {
```
a switch statement with return instead of setting retVal would look nicer here

tomereli replied:
No need to add the legacy modes - just check for the ones we do support (controller, controller and agent, agent)

simonbaren replied:
it has nothing to do with controller/agent.
This API is operating mode (GW,Repeater,IRE....)

simonbaren replied:
Can't switch a string

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -124,78 +124,78 @@ int bpl_cfg_get_passive_mode() { return 0; }
 int bpl_cfg_get_client_roaming() { return 0; }
 
-int bpl_cfg_get_device_info(BPL_DEVICE_INFO *device_info) { return -1; }
+int bpl_cfg_get_device_info(BPL_DEVICE_INFO *device_info) { return RETURN_ERR; }
```
commit message missing sign off by

tomereli commented on framework/platform/bpl/linux/platform_db:
```diff
@@ -0,0 +1,2 @@
+management_mode=Multi-AP-Controller-and-Agent
+operating_mode=Gateway
```
add new line at the end of file

tomereli replied:
also configure vscode to do that automatically - https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files

tomereli reviewed (approved):
minor comments, unless @arnout has other comments you can fix whats left, autosquash and merge.

Commit:
tools: docker: use prplmesh_utils.sh in scripts

Update start-controller, start-controller-agent and start-agent startup
scripts to use prplmesh_utils.sh.
Also align entrypoint.sh to interface name changes.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: config: change default wired iface to eth0

In docker containers we are using, the wired interface is eth0, not
eth1, so change the default in linux platform appropriately.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
agent: backhaul fsm: skip dhcp

Skip dhcp state in backhaul FSM since it is not needed for any platorm.
It is expected that platform initialization will either run dhcp client
on the bridge continuously or configure static IP.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
scripts: prplmesh_utils.sh platform_db support

Add platform_db automatic creation by prplmesh_utils.sh according to the
mode in which the script is invoked, and write that to
/tmp/prplmesh_platform_db.

While we're at it, change eth0 to sim-eth0 since if
prplmesh_utils.sh runs on the host, it will mess up networking.
Using a ~unique name for the dummy interface is good practice anyhow,
and allows us to run the same script in both containers and hosts.
Changing wlan0/2 to sim-wlan0/2 will be done in a separate commit.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on framework/platform/bpl/CMakeLists.txt:
```diff
@@ -56,6 +56,9 @@ elseif (TARGET_PLATFORM STREQUAL "linux")
 
     file(GLOB_RECURSE bpl_platform_sources ${MODULE_PATH}/linux/*.c*)
 
+    add_definitions(-DPLATFORM_DB_PATH="${CMAKE_INSTALL_PREFIX}/share/prplmesh_platform_db")
```
In general, I think it's better to do this update together with the change that actually uses that definition. However, it's not needed to rework just for that. :ok:

arnout commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,17 +15,124 @@ using namespace mapf;
+        if (!in_conf_file.is_open()) {
+            MAPF_ERR("Failed oppening file " << PLATFORM_DB_PATH);
+            return RETURN_ERR;
+        }
```
This pattern of trying to open a file that may be in multiple locations is used in several places. It would be useful to refactor -> #90. :ok:

arnout commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,17 +15,124 @@ using namespace mapf;
+    }
+
+    while (std::getline(in_conf_file, line)) {
+        utils::trim(line);
```
It would be simpler to first chop off the comment, then trim, then check empty - that way, you only need to handle the comment once. :ok:

arnout commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,17 +15,124 @@ using namespace mapf;
+        if (line.empty()) continue; // Empty line
+        if (line.at(0) == '#') continue; // Commented line
+        if (line.compare(0, param.size(), param) != 0) continue; // Not the param we look for
+        
```
To avoid that `param` has to contain the `=`, you can do:

```
line.erase(0, param.size());
utils::ltrim(line);
if (line.at(0) != '=') continue; // Longer param name
utils::ltrim(line, "=");
```

This will also allow spaces around the `=`.

However, all of this is not very important, so :ok: .

arnout commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,17 +15,124 @@ using namespace mapf;
+        return RETURN_ERR;
+    }
+
+    value = utils::stoi(str_value);
```
Meh, `stoi` doesn't do error handling, that's bad... :ok:

tomereli replied:
why? it is implemented with error handling in `beerocks_string_utils.cpp`

arnout replied:
It doesn't handle errors, it hides them and returns 0. When you need an integer value and it's non-integer in the config file, some kind of failure should be flagged.

But it's not terribly important.

arnout commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -15,17 +15,124 @@ using namespace mapf;
+
+    if (mode_str == "Multi-AP-Controller-and-Agent") 
+        return 1;
+    else if (mode_str == "Multi-AP-Agent")
```
Can we support "Multi-AP-Controller" as well? :ok:

tomereli replied:
Yes, will add it.

arnout commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -144,78 +144,78 @@ int bpl_cfg_get_passive_mode() { return 0; }
 int bpl_cfg_get_client_roaming() { return 0; }
 
-int bpl_cfg_get_device_info(BPL_DEVICE_INFO *device_info) { return -1; }
+int bpl_cfg_get_device_info(BPL_DEVICE_INFO *device_info) { return RETURN_ERR; }
```
TBH I think this change is quite useless...

First of all, returning `int` is inconsistent with the rest of the code; `bool` would be more appropriate.

But if we return `int`, it's especially in the callers that it's important that we don't use a magic value. For the functions themselves, you have the rest of the file to refer to.

And for sure, the magic values should be defined in the bpl header itself, not in the rdkb and linux specific code. But then, it should be defined as an enum, not as defines.

That said, this change is definitely making things better and not worse, and it has been done now, so :ok: .

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -613,7 +613,7 @@ bool main_thread::backhaul_fsm_main(bool &skip_select)
             if (m_sConfig.eType == SBackhaulConfig::EType::Wireless) {
```
Typo in commit message: continuesly -> continuously.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -613,7 +613,7 @@ bool main_thread::backhaul_fsm_main(bool &skip_select)
                 FSM_MOVE_STATE(INIT_HAL);
             } else { // EType::Wired
-                FSM_MOVE_STATE(WIRED_BRIDGE_DHCP);
+                FSM_MOVE_STATE(MASTER_DISCOVERY);
```
We should also remove the WIRED_BRIDGE_DHCP and WIRED_BRIDGE_DHCP_WAIT states entirely.

That can be done as a follow-up commit, so I'm not going to make it block approval. However, please don't forget.

tomereli replied:
Better do it now while i'm at it - will push a fixup

arnout commented on agent/CMakeLists.txt:
```diff
@@ -34,7 +34,7 @@ elseif (TARGET_PLATFORM STREQUAL "linux")
     set(BEEROCKS_NOTIFY 0)
     set(BEEROCKS_BRIDGE_IFACE "br-lan")
-    set(BEEROCKS_BH_WIRE_IFACE "eth0")
+    set(BEEROCKS_BH_WIRE_IFACE "prplmesh-eth0")
```
NIC names are (in some cases) limited to 11 characters, so this will be truncated to `prplmesh-et`. So `prplmesh` must be abbreviated. @wcloetens suggested `sim0` on Slack, but I'd make it `sim-eth0`.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -29,30 +29,42 @@ killall_program() {
+    management_mode=${1-Multi-AP-Controller-and-Agent}
+    operating_mode=${2-Gateway}
+
+    mkdir -p /tmp/@PLATFORM_DB_PATH_TEMP@/
```
I must have missed something somewhere... `PLATFORM_DB_PATH_TEMP` seems to be only defined as a cpp macro, not as a CMake variable. Does this work? Doesn't this get substituted with just empty?

Also, the defintiion of the cpp macro already includes the /tmp part.

tomereli replied:
We call it a "foox" in hebrew - meaning it is pure luck :)
@PLATFORM_DB_PATH_TEMP@ is indeed substituted to empty, so the directory is just /tmp which is why it works. I'll fix it.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -29,30 +29,42 @@ killall_program() {
+        echo -e "management_mode=${management_mode}"
+        echo -e "operating_mode=${operating_mode}"
+        echo -e "wired_backhaul=1"
+    } > /tmp/@PLATFORM_DB_PATH_TEMP@/prplmesh_platform_db
```
It *happens* to work if `@PLATFORM_DB_PATH_TEMP@` is substituted by nothing, but I don't think it's intended this way...

arnout commented on tools/docker/runner/entrypoint.sh:
```diff
@@ -8,10 +8,15 @@ run() {
 run ifconfig eth0 0.0.0.0
+run ifconfig prplmesh-eth0 up
+run ifconfig wlan0 up
+run ifconfig wlan2 up
```
Why are these three suddenly needed? Isn't the `ip link up` that is done a bit higher sufficient?

tomereli replied:
no, ip link add doesn't actually support adding an interface and bringing it up in a single command.

tomereli replied:
but I will change it all to use ip command.

arnout commented on tools/docker/runner/start-agent:
```diff
@@ -1,3 +1,3 @@
-${INSTALL_DIR}/bin/beerocks_agent
+#!/bin/sh
+${INSTALL_DIR}/scripts/prplmesh_utils.sh start -s -m a
+tail -f /dev/null # hack so the script will not exit forcing the container to stop
```
It would also make sense to run the CLI as a foreground process instead. :ok:

tomereli replied:
The agent doen't have a CLI

arnout reviewed (approved):
As usual, most of my comments are optional; I added :ok: to make that clear.

There are still two essential changes: prplmesh-eth0 is too long, and @PLATFORM_DB_PATH_TEMP@ doesn't get replaced (and maybe doesn't need to). Since both are rather trivial to fix, I'm approving already - but the intention is to fix them, of course :-)
